### PR TITLE
Update WebRTC Storage Canary

### PR DESCRIPTION
**Pull request by stefankiesz**
## Overview
WebRTC Storage Canary Updates:
- Revamps consumer-end Canary
  - Improves code modularity
  - Adds `TimeToFirstFragment` metric
- Updates webrtc-end Canary to align with upcoming WebRTC-C Storage SDK release
  - `JoinStorageSession` no longer called within peer-side app
  - Removes use of any customer-facing MediaServer-related API
## Consumer-End Changes
- Separates the two consumer-end metric calculations `FragmentContinuity` and `TimeToFirstFragment` into individual methods which are called based on the configured CANARY_LABEL
- Moves CloudWatch metric publishing from within the calculation methods to a designated method
- Fixes Consumer Canary issue of relying on `System.exit(0)` to exit the application
  - Discovered that active thread count was not being decremented after calling `PutMetricDataRequest` to CloudWatch, perhaps due to a thread completion indicator not getting emitted after the Timer Task increments to next iteration
  - CloudWatch client is switched from async version to synchronous to solve this
## WebRTC-End Changes
- Removes all changes made to `Peer` class
- Removes `useMediaStorage` flag from Canary as it is not longer used by `Peer`
- Introduces `isStorage` configurable parameter to replace `useMediaStorage` for CloudWatch metric-labeling purposes
- Removes the termination of Canary upon disconnection from peer to allow for automatic reconnect after the 60min connection limit
### Consumer-End
Set env vars:
```
export CANARY_STREAM_NAME=<storage-configured streamName>
export CANARY_DURATION_IN_SECONDS=<300 or more for Media Server compatibility>
export CANARY_LABEL=<WebrtcLongRunning or WebrtcPeriodic>
export AWS_DEFAULT_REGION=us-west-2
```
Run:
```
java -classpath target/aws-kinesisvideo-producer-sdk-canary-consumer-1.0-SNAPSHOT.jar:$(cat tmp_jar) -Daws.accessKeyId=${AWS_ACCESS_KEY_ID} -Daws.secretKey=${AWS_SECRET_ACCESS_KEY} com.amazon.kinesis.video.canary.consumer.WebrtcStorageCanaryConsumer
```

### WebRTC-End
Set env vars:
```
export CANARY_CHANNEL_NAME=<storage-configured channelName>
export CANARY_IS_MASTER=TRUE
export CANARY_TRICKLE_ICE=TRUE
export AWS_KVS_LOG_LEVEL=1
export CANARY_DURATION_IN_SECONDS=<300 or more for Media Server compatibility>
export CANARY_LABEL=<WebrtcLongRunning or WebrtcPeriodic>
```
Run:
```
./kvsWebrtcCanaryWebrtcStorage
```

<br>
<br>
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.ArrayList;
+
+
+/*
+    CanaryFragmentList allows for intervalMetricsTask calls to access the same data by passing a CanaryFragmentList object.
+ */
+
+public class CanaryFragmentList {
+
+    private List<CanaryFragment> fragmentList = new ArrayList<>();
```
Comment (disa6302) [inline canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java:13]: As a convention, it is preferable to prefix the class variable names with m. Can we have this change throughout wherever relevant?
`fragmentList` -> `mFragmentList`

File: canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    public CanaryFragmentList() {
+    }
+
+    public void setFragmentList(List<CanaryFragment> fragmentList) {
+        this.fragmentList = fragmentList;
+    }
+
+    public List<CanaryFragment> getFragmentList() {
+        return fragmentList;
```
Comment (disa6302) [inline canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java:24]: Nit: `this.`

File: canary/webrtc-c/jobs/runner.groovy [hunk 3]
```diff
@@ -12,14 +12,12 @@ CREDENTIALS = [
     ]
 ]
 
-def buildProject(useMbedTLS, thing_prefix) {
-    echo 'Flag set to ' + useMbedTLS
+def buildWebRTCProject(useMbedTLS, thing_prefix) {
     checkout([$class: 'GitSCM', branches: [[name: params.GIT_HASH ]],
               userRemoteConfigs: [[url: params.GIT_URL]]])
 
-    def configureCmd = "cmake .. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS_DEBUG=\"-g -O0\" -DCMAKE_INSTALL_PREFIX=\"\$PWD\""
```
Comment (disa6302) [inline canary/webrtc-c/jobs/runner.groovy:20]: Hmm. I see some changes from main branch have got reverted? Can we confirm that this branch is rebased with master?

File: canary/webrtc-c/jobs/runner.groovy [hunk 4]
```diff
@@ -254,24 +353,27 @@ pipeline {
...
                     job: env.JOB_NAME,
+
                     parameters: [
+                      string(name: 'AWS_DEFAULT_REGION', value: params.AWS_DEFAULT_REGION),
                       string(name: 'AWS_KVS_LOG_LEVEL', value: params.AWS_KVS_LOG_LEVEL),
                       booleanParam(name: 'IS_SIGNALING', value: params.IS_SIGNALING),
+                      booleanParam(name: 'IS_STORAGE', value: params.IS_STORAGE),
                       booleanParam(name: 'USE_TURN', value: params.USE_TURN),
                       booleanParam(name: 'USE_IOT', value: params.USE_IOT),
-                      booleanParam(name: 'IS_PROFILING', value: params.IS_PROFILING),
```
Comment (disa6302) [inline canary/webrtc-c/jobs/runner.groovy:262]: Same here. 
I think when you rebased, you only took the PR branch changes and removed master branch changes. It should have been the other way round to include the master branch changes in your branch.

File: canary/webrtc-c/jobs/orchestrator.groovy [hunk 5]
```diff
@@ -4,10 +4,13 @@ RUNNER_JOB_NAME_PREFIX = "webrtc-canary-runner"
...
 PERIODIC_PROFILING_DURATION_IN_SECONDS = 90
 LONG_RUNNING_DURATION_IN_SECONDS = 0
+DURATION_IN_SECONDS_5_MIN = 300
+DURATION_IN_SECONDS_45_MIN = 2700
+DURATION_IN_SECONDS_65_MIN = 3900
 MIN_RETRY_DELAY_IN_SECONDS = 60
 COLD_STARTUP_DELAY_IN_SECONDS = 60 * 60
 GIT_URL = 'https://github.com/aws-samples/amazon-kinesis-video-streams-demos.git'
-GIT_HASH = 'master'
+GIT_HASH = 'newer-ingestion-canary' // TODO: Revert to 'master'
```
Comment (stefankiesz) [inline canary/webrtc-c/jobs/orchestrator.groovy:13]: Reminder to self to revert this back to master branch

File: canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java (hunk 1 above)
Comment (stefankiesz) [reply canary/consumer-java/src/main/java/com/amazon/kinesis/video/canary/consumer/CanaryFragmentList.java]: Updated. I didn't do this for CanaryListFragmentWorker as I'm trying to keep the code as similar to the example it's modeled after as possible.

**Comment by stefankiesz** [comment]
These changes will be included in the sample canary merge.

**Pull request closed without merging**